### PR TITLE
[MRG+1] MAINT deprecated 'spectral' in favor of 'nipy_spectral'

### LINKED ISSUE #7315: Spectral vs spectral Deprecation warning
This issue has [already been opened](https://github.com/matplotlib/matplotlib/issues/879) : there are 2 colormap with almost identical name: spectral and Spectral (casing counts). Name was [changed](https://github.com/matplotlib/matplotlib/pull/921)  to [`nipy_spectral`](https://github.com/matplotlib/matplotlib/pull/1127) but there is no warning which leads to @NelleV complaining and engaged a couple of coworkers in long bughunt. 

I would suggest that the use of spectral (lower case) raised a deprecation warning or even raise an error as default are being changed.

### PR DESCRIPTION
**Pull request by NelleV**
This raises a deprecation warning for the deprecated "spectral" colormap.
I think I covered all the possible calls to "spectral".

closes #7315

### CODE REVIEW

**Comment by NelleV** [comment]
@Carreau Do you want to review? :)

**Comment by Carreau** [comment]
> @Carreau Do you want to review? :)

Yes :-) The name vas changed in 2012, so I would not be shy as marking as deprecated since at least matplotlib 1.5.

You might want to warn on `spectral_r` as well (constructed in `lib/matplotlib/cm.py`)

For post 2.0 / Subsequent PRs:

- It might be hard, but add a warning filter that turns use of spectral during testing (and doc building ?) into an error.
- Have `keys()`, `iterkeys()`, `items()` not list spectral and `spectral_r`

An easy implementation would be to actually remove the data find the dict, and if name==spectral, return super()__getattra__('nipy_spectral_'). But that can be done later.

**Comment by Carreau** [comment]
`doc/users/image_tutorial.rst` uses `spectral`...

**Comment by NelleV** [comment]
Yeah, I forgot to look at the documentation and "fix" it…

**Comment by Carreau** [comment]
Sorry I'm slow to review I'm in the train, tethering and the car is shaking quite a bit.

**Comment by NelleV** [comment]
Updated it to deprecate `spectral_r` as well.

**Comment by efiring** [comment]
Looks OK to me.

File: lib/matplotlib/_cm.py [hunk 1]
```diff
@@ -1365,7 +1366,19 @@ def gfunc32(x):
...
-datad = {
+class _deprecation_datad(dict):
+    def __getitem__(self, key):
+        if key in ["spectral", "spectral_r"]:
+            warn_deprecated(
+                "2.0",
+                name="spectral",
+                alternative="nipy_spectral",
+                obj_type="colormap"
+                )
```
Comment (Carreau) [inline lib/matplotlib/_cm.py]: do you want to make the message dynamic for `spectral_r` as well ? I can see pros and cons to both.

File: lib/matplotlib/cm.py [hunk 2]
```diff
@@ -16,11 +16,11 @@
 import matplotlib as mpl
 import matplotlib.colors as colors
 import matplotlib.cbook as cbook
-from matplotlib._cm import datad
+from matplotlib._cm import datad, _deprecation_datad
 from matplotlib._cm import cubehelix
 from matplotlib._cm_listed import cmaps as cmaps_listed
 
-cmap_d = dict()
+cmap_d = _deprecation_datad()
```
Comment (Carreau) [inline lib/matplotlib/cm.py]: Note that line 88 of this file will likely trigger the deprecation warning if I'm not wrong, as it loop through all the colormap to build the reversed ones. Not sure what we can do about the though. Just pointing it out.

**Comment by Carreau** [comment]
> Updated it to deprecate spectral_r as well.

+1. Thanks @NelleV !

File: lib/matplotlib/_cm.py (hunk 1 above)
Comment (NelleV) [inline lib/matplotlib/_cm.py]: I was going to be lazy, and not do it, but… maybe I should.

**Comment by NelleV** [comment]
I'm now catching warnings when creating the reverse colormaps to avoid raising the deprecation warning when cm.py is imported (ie, when anyone uses matplotlib).

**Comment by tacaswell** [comment]
Do we want to also deprecate `spectral` function in `pyplot` as well?

**Comment by NelleV** [comment]
Yes, we should… but how to do this with the function being generated?

**Comment by tacaswell** [comment]
https://github.com/matplotlib/matplotlib/blob/v2.x/boilerplate.py#L333 split the list into two (deprecated and not) and make a new template which includes the deprecation warning.

**Comment by NelleV** [comment]
Got it. I'll do that.

**Comment by NelleV** [comment]
Should I run boilerplate and commit the function as well or should we wait for the release?

**Comment by efiring** [comment]
Yes, it's best to keep pyplot in sync with everything else.

**Comment by QuLogic** [comment]
Wow, `pyplot.py` seems to be pretty outdated; will have to be careful when backporting.

**Comment by tacaswell** [comment]
Should probably just re-run boilerplate after doing the backport.

**Comment by efiring** [comment]
backported as f87100a (including the pyplot.py regenerated by boilerplate.py)

**Pull request merged**